examples/gno.land/p/demo/paginate/paginate.gno

```diff
@@ -37,60 +103,69 @@ func RenderTreeByPage(
 	}
 
 	str := ""
-	numpages := (t.Size() / size) + 1
+	numPages := (t.Size() / size)
+	if numPages < 1 {
+		return ""
+	}
 
 	// render pagination
 	if pageFn == nil {
-		// do not render pagination
-	} else if page > numpages {
+		// render tree items
+		var offset int
+		if reverse {
+			offset = (numPages - page) * size
+		} else {
+			offset = (page - 1) * size
+		}
+		t.IterateByOffset(offset, size,
+			func(key string, value interface{}) bool {
+				r := value.(renderer)
+				str += "\n"
+				str += r.Render()
+				str += "\n"
+				str += "\n"
+
+				return true
+			},
+		)
+		return str
+	} else if page > numPages {
 		// page out of bounds.
 		// [1](...) [2](...) ... [101](...) [102)(...)
 		// where ... is result of pageFn().
-		// TODO: improve this function by renderring
-		// only the first and last few pages. XXX
-		for p := 1; p <= numPages; p++ {
-			pageurl := pageFn(p, size, reverse)
-			str += "[" + itoa(page) + "](" +
-				escMDURL(pageurl) + ") "
-		}
+		str += navigation(numPages, page, reverse, pageFn)
 		str += "\n"
-		str += "page " + itoa(page) + " out of bounds\n"
+		str += "page " + strconv.Itoa(page) + " out of bounds\n"
 		return str
-	}
-	// [1](...) *2* [3](...) [4](...) ... [101](...) [102)(...)
-	// where ... is result of pageFn().
-	// TODO: improve this function by only printing the first few,
-	// the last few, and the ones surrounding the current page.
-	// Must make this improvement to be strictly "memory correct".
-	if numpages > 0 {
-		for p := 1; p <= numPages; p++ {
-			pageurl := pageFn(p, size, reverse)
-			str += "[" + itoa(page) + "](" +
-				escMDURL(pageurl) + ") "
-		}
+	} else {
+		// [1](...) *2* [3](...) [4](...) ... [101](...) [102)(...)
+		// where ... is result of pageFn().
+		str += navigation(numPages, page, reverse, pageFn)
 		str += "\n"
-	}
 
-	// render tree items
-	offset := (page - 1) * size
-	t.IterateByOffset(offset, size,
-		func(key string, value interface{}) bool {
-			var r = value.(renderer)
-			str += r.Render()
-			str += "\n"
-			str += "\n"
-		},
-	)
-
-	// render pagination again if there are pages.
-	if numpages > 0 {
-		for p := 1; p <= numPages; p++ {
-			pageurl := pageFn(p, size, reverse)
-			str += "[" + itoa(page) + "](" +
-				escMDURL(pageurl) + ") "
+		// render tree items
+		var offset int
+		if reverse {
+			offset = (numPages - page) * size
+		} else {
+			offset = (page - 1) * size
 		}
-		str += "\n"
+		t.IterateByOffset(offset, size,
+			func(key string, value interface{}) bool {
+				r := value.(renderer)
+				str += "\n"
+				str += r.Render()
+				str += "\n"
+				str += "\n"
+
+				return true
+			},
+		)
 	}
 
+	// render pagination again if there are pages.
+	str += navigation(numPages, page, reverse, pageFn)
```

Comment:
You should re-use the last call to `navigation()` to save some CPU cycles.

Reply:
:+1: